### ISSUE TITLE
[DocDB] Consider replacing fsync with fdatasync on the WAL write path

### ISSUE THREAD

**Issue by bmatican** [yugabyte/yugabyte-db#16598]
### Description

As per discussions with @fritshoogland-yugabyte , given we pre-allocate our WAL files, we might be able to relieve some disk pressure by using fdatasync instead of fsync
- for any calls during the pre-allocated part of the file, we should see benefits
- for the final call, when we cross the pre-allocated section of the file (and thus will close the file afterwards), fdatasync anyway seems to "promote" to fsync, thus being functionally what we want

Frits is working on a good test bed for us to validate the win from this, but implementation wise, this should be relatively easy to pull off and put behind a new gflag.
FYI @rthallamko3 @Huqicheng @yusong-yan @ttyusupov 

### Warning: Please confirm that this issue does not contain any sensitive information

- [X] I confirm this issue does not contain any sensitive information.

**Comment by fritshoogland-yugabyte** [comment]
In general, fsync forces the OS to flush the inode dirty pages as well as the filesystem journal. For me info see: https://dev.to/yugabyte/the-anatomy-of-xfs-fsync-4ael

There is an fsync call that, if safe, only flushes the dirty pages and not the journal, which is "fdatasync".
The journal can be a serialisation point, which is made clear in the above text about fsync.
The fdatasync is safe, because if it detect a file structure change that needs flushing of the journal, it will automatically perform that. 

PostgreSQL also uses fdatasync. The fsync options and implementation have been thoroughly been looked at in 2018 when linux was found to optionally not show IO errors in some cases, which has been fixed.

I performed a simple test in a VM on my laptop to see the difference between calling fsync() and fdatasync() after a write() call:

[Fsync Summary - Criterion.rs.pdf](https://github.com/yugabyte/yugabyte-db/files/11104519/Fsync.Summary.-.Criterion.rs.pdf)

In short, this shows the difference between fsync and fdatasync is roughly 12ms and 6ms.
